draft-welzl-iccrg-pacing.md

```diff
@@ -70,11 +70,43 @@ Because of its known advantages, pacing has become common in implementations of
 
 # Pacing: general considerations and consequences
 
-TODO
+## More likely to saturate a bottleneck {#losstypes}
+
+We can distinguish between two reasons for packet losses that are due to congestion at a bottleneck with a DropTail (FIFO) queue:
+
+1. A flight of N packets arrives. The amount of data in this flight exceeds the amount of data that can be transmitted by the bottleneck during the flight's arrival plus the queue length, i.e. some data do not fit into the queue.
+2. The bottleneck is fully saturated. The queue is full, and packets drain from it more slowly than new packets arrive.
+
+The second type of loss matches the typical expectation of a congestion control algorithm: the cwnd value when loss happens is indicative of the bottleneck being fully saturated. When the first type of loss happens, however, a sender's cwnd can be much smaller than the Bandwidth*Delay Product (BDP) of the path (the amount of data that can be in flight, ignoring the queue). In the absence of other traffic, the probability for the first type of loss to happen depends on the queue length and the ratio between the departure and the arrival rate during the flight's arrival. By introducing time gaps between the packets of a burst, this ratio is increased, i.e. the difference between the departure and the arrival rate becomes smaller, and the second type of loss is more likely.
+
+For example, consider a network path with a bottleneck capacity of 50 Mbit/s, a queue length of 15000 bytes (or 10 packets of size 1500 bytes) and an RTT of 30 ms. Assume that all packets emitted by the sender have a size of 1500 bytes. Then, the BDP equals 125 packets. The bottleneck of this network path is fully saturated when a (BDP + queue length) amount of bytes are in flight: 135 packets.
+
+In this network, the first type of loss can happen as follows: say, N=40 packets arrive at this bottleneck at a rate of 100 Mbit/s. In an otherwise empty network and assuming an initial window of 10 packets and no delayed ACKs, this occurs in the third round of slow start without pacing, provided that the capacities of all links before the bottleneck are at least 100 Mbit/s.
+In this case, an overshoot will occur: packets are forwarded with half their arrival rate, i.e. less than 20 packets can be forwarded during the burst's arrival. The remaining 20 (or more) packets cannot fit into the 10-packet queue. A cwnd of 40 packets is much smaller than the (BDP + queue) limit of 135 packets, and the bottleneck is not fully saturated.
+
+Let us now assume that the flight of 40 packets is instead paced, such that the arrival rate only mildly exceeds the departure rate -- e.g., they arrive at a rate of 60 Mbit/s. When the last packet of this flight arrives at the bottleneck, 5/6 * 39 = 32.5 packets should already have been transferred. Since only complete packets are sent, 32 packets have really been transferred, and the remaining 40-32 = 8 packets fit in the queue. No loss occurs.
```

Comment:
Further, should we say that pacing gives more time for the growth of the queue to be detected by the transport (e.g. within a delay signal, or in the case of AQM via ECN) prior to it resulting in necessary loss, in comparison to the non-pacing case?

Reply:
Hm, does this really only play out positively? It may depend on the AQM...  for creating a more sustained queue growth, a paced flow may end up being "punished" earlier? Or causing punishment for all flows earlier?

Overall I personally think that everything to do with pacing is good, because it makes the behavior more in line with the expectations of a CC algorithm; but in specific scenarios, given how algorithms behave today, bad things MAY happen, and I'd like to avoid a confusion... so let's be a bit cautious with what we say.  The stuff I do say here, based on FIFO and a single flow, I'm very sure about.

Reply:
I added discussing dynamics with AQM as a todo item in issue #3

Reply:
Agreed!